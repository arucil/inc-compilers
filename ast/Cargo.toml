[package]
name = "ast"
version = "0.1.0"
edition = "2018"

[dependencies]
pretty = "0.10.0"
support = { path = "../support" }

[dev-dependencies]
pretty_assertions = "0.7.2"
