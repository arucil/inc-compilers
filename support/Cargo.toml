[package]
name = "support"
version = "0.1.0"
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
assert_cmd = "2.0.4"
predicates = "2.1.1"
pretty_assertions = "1.2.1"
scopeguard = "1.1.0"
