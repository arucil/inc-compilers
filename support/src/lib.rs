pub mod compile_error;
pub mod range;
pub mod test_cli;

pub use self::compile_error::*;
pub use self::range::*;
pub use self::test_cli::*;
