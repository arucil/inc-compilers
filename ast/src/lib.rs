#![feature(never_type, box_syntax)]

use id_arena::{Arena, Id};
use pretty::*;
use std::fmt::{self, Debug, Formatter};
use std::iter;
use support::Range;

pub mod parser;

pub use parser::{parse, Result};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Program<VAR = String, TYPE = ()> {
  pub body: Vec<Exp<VAR, TYPE>>,
  pub types: Arena<CompType>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Exp<VAR = String, TYPE = ()> {
  pub kind: ExpKind<VAR, TYPE>,
  pub range: Range,
  pub ty: TYPE,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ExpKind<VAR, TYPE> {
  Int(i64),
  Prim {
    op: (Range, &'static str),
    args: Vec<Exp<VAR, TYPE>>,
  },
  Var(VAR),
  Str(String),
  Let {
    var: (Range, VAR),
    init: Box<Exp<VAR, TYPE>>,
    body: Box<Exp<VAR, TYPE>>,
  },
  Bool(bool),
  If {
    cond: Box<Exp<VAR, TYPE>>,
    conseq: Box<Exp<VAR, TYPE>>,
    alt: Box<Exp<VAR, TYPE>>,
  },
  Set {
    var: (Range, VAR),
    exp: Box<Exp<VAR, TYPE>>,
  },
  Get(VAR),
  Begin {
    seq: Vec<Exp<VAR, TYPE>>,
    last: Box<Exp<VAR, TYPE>>,
  },
  While {
    cond: Box<Exp<VAR, TYPE>>,
    body: Box<Exp<VAR, TYPE>>,
  },
  Void,
  Print(
    /// nonempty
    Vec<Exp<VAR, TYPE>>,
  ),
  NewLine,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Type {
  Int,
  Bool,
  Str,
  Comp(TypeId),
  Void,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct TypeId(Id<CompType>);

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum CompType {
  Vector(Vec<Type>),
}

#[derive(Clone, PartialEq, Eq, Hash)]
pub struct IdxVar {
  pub name: String,
  pub index: usize,
}

impl IdxVar {
  pub fn new<S: ToString>(name: S) -> Self {
    let name = name.to_string();
    if let Some(i) = name.find('.') {
      if let Ok(index) = name[i + 1..].parse::<usize>() {
        return Self {
          name: name[..i].to_owned(),
          index,
        };
      }
    }
    Self { name, index: 0 }
  }
}

impl Debug for IdxVar {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    write!(f, "{}.{}", self.name, self.index)
  }
}

impl<VAR: Debug, TYPE: Debug> Program<VAR, TYPE> {
  pub fn to_string_pretty(&self) -> String {
    let doc = self.to_doc();
    let mut buf = String::new();
    doc.render_fmt(30, &mut buf).unwrap();
    buf
  }

  fn to_doc(&self) -> RcDoc {
    RcDoc::intersperse(self.body.iter().map(|exp| exp.to_doc()), Doc::line())
  }
}

impl<VAR: Debug, TYPE: Debug> Exp<VAR, TYPE> {
  fn to_doc(&self) -> RcDoc {
    match &self.kind {
      ExpKind::Int(n) => RcDoc::text(format!("{}", n)),
      ExpKind::Str(s) => RcDoc::text(format!("{:?}", s)),
      ExpKind::Var(var) | ExpKind::Get(var) => {
        RcDoc::text(format!("{:?}", var))
      }
      ExpKind::Prim { op, args } => RcDoc::text("(")
        .append(
          RcDoc::intersperse(
            iter::once(RcDoc::text(op.1))
              .chain(args.iter().map(|arg| arg.to_doc())),
            Doc::line(),
          )
          .nest(1)
          .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::Let { var, init, body } => RcDoc::text("(")
        .append(
          RcDoc::text("let")
            .append(Doc::line())
            .append(
              RcDoc::text("[")
                .append(
                  RcDoc::text(format!("{:?}", var.1))
                    .append(Doc::line())
                    .append(init.to_doc())
                    .nest(1)
                    .group(),
                )
                .append(RcDoc::text("]")),
            )
            .append(Doc::line())
            .append(body.to_doc())
            .nest(1)
            .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::Bool(true) => RcDoc::text("#t"),
      ExpKind::Bool(false) => RcDoc::text("#f"),
      ExpKind::If { cond, conseq, alt } => RcDoc::text("(")
        .append(
          RcDoc::text("if")
            .append(Doc::line())
            .append(cond.to_doc())
            .group()
            .append(Doc::line())
            .append(conseq.to_doc().append(Doc::line()).append(alt.to_doc()))
            .nest(1)
            .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::Set { var, exp } => RcDoc::text("(")
        .append(
          RcDoc::text("set!")
            .append(Doc::line())
            .append(RcDoc::text(format!("{:?}", var.1)))
            .append(Doc::line())
            .append(exp.to_doc())
            .nest(1)
            .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::Begin { seq, last } => RcDoc::text("(")
        .append(
          RcDoc::text("begin").append(Doc::line()).append(
            RcDoc::intersperse(
              seq
                .iter()
                .chain(std::iter::once(&**last))
                .map(|exp| exp.to_doc()),
              Doc::line(),
            )
            .nest(1)
            .group(),
          ),
        )
        .append(RcDoc::text(")")),
      ExpKind::While { cond, body } => RcDoc::text("(")
        .append(
          RcDoc::text("while")
            .append(Doc::line())
            .append(cond.to_doc())
            .group()
            .append(Doc::line())
            .append(body.to_doc())
            .nest(1)
            .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::Void => RcDoc::text("(void)"),
      ExpKind::Print(args) => RcDoc::text("(")
        .append(
          RcDoc::intersperse(
            iter::once(RcDoc::text("print"))
              .chain(args.iter().map(|arg| arg.to_doc())),
            Doc::line(),
          )
          .nest(1)
          .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::NewLine => RcDoc::text("(newline)"),
    }
  }
}
