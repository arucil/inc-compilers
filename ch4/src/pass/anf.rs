use ast::{Exp, IdxVar, Program};
use support::Range;

pub fn anf(prog: Program<IdxVar>) -> Program<IdxVar> {
  let mut counter = 0;
  Program {
    body: prog
      .body
      .into_iter()
      .map(|(range, exp)| (range, anf_exp(range, exp, &mut counter)))
      .collect(),
  }
}

fn anf_exp(range: Range, exp: Exp<IdxVar>, counter: &mut usize) -> Exp<IdxVar> {
  match exp {
    Exp::Let { var, init, body } => Exp::Let {
      var,
      init: box (init.0, anf_exp(init.0, init.1, counter)),
      body: box (body.0, anf_exp(body.0, body.1, counter)),
    },
    Exp::Prim { op, args } => {
      let mut tmps = vec![];
      let args = args
        .into_iter()
        .map(|(range, arg)| (range, atom_exp(range, arg, &mut tmps, counter)))
        .collect();
      tmps
        .into_iter()
        .rfold(
          (range, Exp::Prim { op, args }),
          |(body_range, body), (range, var, init)| {
            (
              body_range,
              Exp::Let {
                var: (range, var),
                init: box (range, init),
                body: box (body_range, body),
              },
            )
          },
        )
        .1
    }
    Exp::If { cond, conseq, alt } => Exp::If {
      cond: box (cond.0, anf_exp(cond.0, cond.1, counter)),
      conseq: box (conseq.0, anf_exp(conseq.0, conseq.1, counter)),
      alt: box (alt.0, anf_exp(alt.0, alt.1, counter)),
    },
    Exp::Set { var, exp } => Exp::Set {
      var,
      exp: box (exp.0, anf_exp(exp.0, exp.1, counter)),
    },
    Exp::Begin { seq, last } => Exp::Begin {
      seq: seq
        .into_iter()
        .map(|exp| (exp.0, anf_exp(exp.0, exp.1, counter)))
        .collect(),
      last: box (last.0, anf_exp(last.0, last.1, counter)),
    },
    Exp::While { cond, body } => Exp::While {
      cond: box (cond.0, anf_exp(cond.0, cond.1, counter)),
      body: box (body.0, anf_exp(body.0, body.1, counter)),
    },
    Exp::Print { val, ty } => Exp::Print {
      val: box (val.0, anf_exp(val.0, val.1, counter)),
      ty,
    },
    exp => exp,
  }
}

fn atom_exp(
  range: Range,
  exp: Exp<IdxVar>,
  tmps: &mut Vec<(Range, IdxVar, Exp<IdxVar>)>,
  counter: &mut usize,
) -> Exp<IdxVar> {
  match exp {
    Exp::Prim { op, args } => {
      let args = args
        .into_iter()
        .map(|(range, arg)| (range, atom_exp(range, arg, tmps, counter)))
        .collect();
      let exp = Exp::Prim { op, args };
      assign_var(range, exp, tmps, counter)
    }
    Exp::Let { .. }
    | Exp::If { .. }
    | Exp::Set { .. }
    | Exp::Begin { .. }
    | Exp::While { .. } => {
      let exp = anf_exp(range, exp, counter);
      assign_var(range, exp, tmps, counter)
    }
    exp => exp,
  }
}

fn assign_var(
  range: Range,
  exp: Exp<IdxVar>,
  tmps: &mut Vec<(Range, IdxVar, Exp<IdxVar>)>,
  counter: &mut usize,
) -> Exp<IdxVar> {
  let tmp = IdxVar {
    name: "tmp".to_owned(),
    index: *counter,
  };
  *counter += 1;
  tmps.push((range, tmp.clone(), exp));
  Exp::Var(tmp)
}

#[cfg(test)]
mod tests {
  use super::*;
  use ast::*;
  use insta::assert_snapshot;

  #[test]
  fn nested_prims() {
    let prog =
      parse(r#"(let ([x (read)] [y (+ 2 3)]) (+ (- (read)) (+ y (- 2))))"#)
        .unwrap();
    let prog = super::super::uniquify::uniquify(prog);
    let result = anf(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn init_with_var() {
    let prog = parse(r#"(let ([a 42]) (let ([b a]) b))"#).unwrap();
    let prog = super::super::uniquify::uniquify(prog);
    let result = anf(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn let_in_init() {
    let prog = parse(
      r#"(let ([a (+ (let ([x (read)] [y (- x)]) (+ x (- y))) 7)]) (- a))"#,
    )
    .unwrap();
    let prog = super::super::uniquify::uniquify(prog);
    let result = anf(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn if_form() {
    let prog = parse(
      r#"(+ (- 3 2) (if (not (eq? 2 3)) (- (read)) (> (read) (not #t))))"#,
    )
    .unwrap();
    let prog = super::super::uniquify::uniquify(prog);
    let result = anf(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn begin() {
    let prog = parse(
      r#"
(let ([x0 10])
  (let ([y1 0])
    (+ (+ (begin (set! y1 (read)) x0)
       (begin (set! x0 (read)) y1))
    x0)))"#,
    )
    .unwrap();
    let prog = super::super::uniquify::uniquify(prog);
    let result = anf(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}
