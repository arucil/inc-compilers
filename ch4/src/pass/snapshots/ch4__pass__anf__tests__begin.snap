---
source: ch4/src/pass/anf.rs
expression: result.to_string_pretty()

---
(let
 [x0.0 10]
 (let
  [y1.1 0]
  (let
   [tmp.0
    (begin
    (set! y1.1 (read)) x0.0)]
   (let
    [tmp.1
     (begin
     (set! x0.0 (read)) y1.1)]
    (let
     [tmp.2 (+ tmp.0 tmp.1)]
     (+ tmp.2 x0.0))))))
